azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py

```diff
@@ -137,60 +121,89 @@ def _on_transition_complete(event_data):
 
         self._create_mqtt_provider()
 
-    def on_enter_connected(self, event_data):
-        if event_data.event.name == "_trig_provider_connect_complete":
-            self.on_transport_connected("connected")
-
-    def on_enter_disconnected(self, event_data):
-        if event_data.event.name == "_trig_provider_disconnect_complete":
-            self.on_transport_disconnected("disconnected")
-
-    def _before_action_notify_publish_complete(self, event_data):
-        logger.info("publish complete:" + str(event_data))
-        logger.info("publish error:" + str(event_data.error))
-        if not event_data.error:
-            self.on_event_sent()
-
-    def _after_action_provider_connect(self, event_data):
+    def _call_provider_connect(self, event_data):
         """
         Call into the provider to connect the transport.
-        This is meant to be called by the state machine as an "after" action
+        This is meant to be called by the state machine as part of a state transition
         """
+        logger.info("Calling provider connect")
         self._mqtt_provider.connect()
 
-    def _after_action_provider_disconnect(self, event_data):
+    def _call_provider_disconnect(self, event_data):
         """
         Call into the provider to disconnect the transport.
-        This is meant to be called by the state machine as an "after" action
+        This is meant to be called by the state machine as part of a state transition
         """
+        logger.info("Calling provider disconnect")
         self._mqtt_provider.disconnect()
 
-    def _before_action_add_event_to_queue(self, event_data):
+    def _on_provider_connect_complete(self):
         """
-        Queue an event for sending later.
-        This is meant to be called by the state machine as an "before" action
+        Callback that is called by the provider when the connection has been established
         """
-        # TODO: throw here if event_data.args[0] is not an event/message
-        self._event_queue.put_nowait(event_data.args[0])
+        logger.info("_on_provider_connect_complete")
+        self._trig_provider_connect_complete()
+
+        if self.on_transport_connected:
+            self.on_transport_connected("connected")
+        callback = self._connect_callback
+        if callback:
+            self._connect_callback = None
+            callback()
```

Comment:
for later: this pattern is akin to having a global state - it's not always bad, but it means we now have properties that have their state "tied" to the state machine but enforced by us. I wonder if there's a way to pass these around rather than store them. 
In any case, it's worth keeping an eye on how this "global state" grows over time. 2 of these is OK, if we start having 5 or 7 properties for which we need to ensure consistency, it becomes tricky.

Reply:
also, nitpick: there seems to be a pattern on how we handle callback - consider helper functions/classes? maybe only if it "grows"?

---
In reply to: [237660851](https://github.com/Azure/azure-iot-sdk-python-preview/pull/26#discussion_r237660851) [](ancestors = 237660851)

Reply:
I agree, this became painfully obvious in the Node SDK.  I'm going to chew on this.  

---
In reply to: [237661215](https://github.com/Azure/azure-iot-sdk-python-preview/pull/26#discussion_r237661215) [](ancestors = 237661215,237660851)